californium-core/src/main/java/org/eclipse/californium/core/network/stack/ResponseTimeoutLayer.java

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Institute for Pervasive Computing, ETH Zurich and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ */
+package org.eclipse.californium.core.network.stack;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.californium.core.coap.MessageObserverAdapter;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.network.Exchange;
+
+/**
+ * Handle response timeout for request. It Should mark request as timeout if responseTimeout expired.
+ * 
+ * @see Request#getResponseTimeout()
+ */
+public class ResponseTimeoutLayer extends AbstractLayer {
+	
+	@Override
+	public void sendRequest(final Exchange exchange,final Request request) {
+		if (request.getResponseTimeout() > 0) {
+			
```

Comment:
I'm not sure, but is this done for each single block request?

(I guess, I have to checkout the branch and do some tests with logging ...)

Reply:
Yes for each single block.

Reply:
Though "onResponse()" seems to be only called at the end of a blockwise transfere, I'm not sure if the resulting behaviour is the intended on. 

The blockwise cleanup (server side) is triggered again by each block (and cancel the old schedule)

This response timeout is set upon the first block request and on the following bock requests, it will fire for the first block request regardless of (slowly) incoming follow up blocks. And the response timeout handling for the follow up block request may fire, because they never get an onResponse.

Reply:
> Though "onResponse()" seems to be only called at the end of a blockwise transfere, I'm not sure if the resulting behaviour is the intended on.

Good catch. For me setResponse should be called on each block request.